### PR TITLE
Base code with futures (fixes #3)

### LINKED ISSUE #3: Add to codebase the TestNG listener, the test-data-supplier
- Fix all violations;
- Add TestNG listeners;
- Add test-data-supplier library (+ IntelliJ plugin);
- Provide some sample data to the existing tests.

### PR DESCRIPTION
**Pull request by mkgerasimenko**
- added gradle quality plugin;
- added TestNG listener;
- added test-data-supplier library.

### CODE REVIEW

File: src/main/java/com/waverleysoftware/model/Panel.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package com.waverleysoftware.model;
+
+/**
+ * A simple model class for user processing.
```
Comment (sskorol) [inline src/main/java/com/waverleysoftware/model/Panel.java]: It doesn't seem to be a class for user processing. ;)

File: src/main/java/com/waverleysoftware/model/Panel.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    public String getMasterCode() {
+        return masterCode;
+    }
+
+    public String getType() {
+        return type;
+    }
+
+    public boolean isHasTouchScreen() {
```
Comment (sskorol) [inline src/main/java/com/waverleysoftware/model/Panel.java]: `isHas` sounds weird. Use the better.

File: src/main/java/com/waverleysoftware/model/Sensor.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package com.waverleysoftware.model;
+
+/**
+ * Author mkgerasimenko.
```
Comment (sskorol) [inline src/main/java/com/waverleysoftware/model/Sensor.java]: Add a more straightforward javadoc. Don't leave a default one.

File: src/main/java/com/waverleysoftware/model/Type.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package com.waverleysoftware.model;
+
+/**
+ * Author mkgerasimenko.
```
Comment (sskorol) [inline src/main/java/com/waverleysoftware/model/Type.java]: Leaving default javadoc is not a good practice.

File: src/test/java/com/waverleysoftware/AccessCodeTests.java [hunk 5]
```diff
@@ -1,32 +1,34 @@
...
 
-import org.testng.annotations.AfterClass;
-import org.testng.annotations.BeforeClass;
+import com.waverleysoftware.model.Panel;
 import org.testng.annotations.Test;
 
+import java.util.logging.Level;
+
+/**
+ * Author mkgerasimenko.
```
Comment (sskorol) [inline src/test/java/com/waverleysoftware/AccessCodeTests.java]: See above comments regarding javadocs.

File: src/test/java/com/waverleysoftware/BaseTest.java [hunk 6]
```diff
@@ -3,14 +3,22 @@
 import org.testng.annotations.AfterClass;
 import org.testng.annotations.BeforeClass;
 
+import java.util.logging.Logger;
+
+/**
+ * Author mkgerasimenko.
```
Comment (sskorol) [inline src/test/java/com/waverleysoftware/BaseTest.java]: See comments about javadocs.

**Review by sskorol — CHANGES_REQUESTED** [review]
@mkgerasimenko check comments please

File: src/main/java/com/waverleysoftware/listemers/SecondTestListener.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+import org.testng.ISuite;
+import org.testng.ISuiteListener;
+
+import java.util.logging.Logger;
+
+@SuppressWarnings("JavadocType")
+public class SecondTestListener implements ISuiteListener {
+
+    private static final Logger LOG = Logger.getLogger(TestListener.class.getName());
```
Comment (sskorol) [inline src/main/java/com/waverleysoftware/listemers/SecondTestListener.java:11]: try replacing common java logger with `@Slf4j` annotation, which is a part of Lombok library

File: src/main/java/com/waverleysoftware/model/Panel.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
+package com.waverleysoftware.model;
+
+/**
+ * A simple model class for panel processing.
+ */
+public class Panel {
```
Comment (sskorol) [inline src/main/java/com/waverleysoftware/model/Panel.java:6]: use Lombok to remove boilerplate for all entities

File: src/test/java/com/waverleysoftware/AccessCodeTests.java [hunk 9]
```diff
@@ -1,32 +1,31 @@
...
 
-    @Test(priority = 3)
-    public void accessCodeShouldBeDeleted() {
-        System.out.println("Access code deleted");
+    @Test(dataProvider = "getPanelData", dataProviderClass = DataSuppliers.class)
+    public void accessCodeShouldBeAdded(final Panel panel) {
+        final String conditionForType = "home";
+        if (conditionForType.equals(panel.getType())
+                && panel.hasTouchScreen()
+                && LOG.isLoggable(Level.INFO)) {
```
Comment (sskorol) [inline src/test/java/com/waverleysoftware/AccessCodeTests.java:18]: with `@Slf4j` you don't need to check of corresponding logging type is available

**Review by sskorol — CHANGES_REQUESTED** [review]
@mkgerasimenko  check comments please

File: src/main/java/com/waverleysoftware/listemers/SecondTestListener.java (hunk 7 above)
Comment (mkgerasimenko) [reply src/main/java/com/waverleysoftware/listemers/SecondTestListener.java:11]: I did it in the  **5-lombok-assert-annotation** branch. Please, look at the changes.

File: src/main/java/com/waverleysoftware/model/Panel.java (hunk 8 above)
Comment (mkgerasimenko) [reply src/main/java/com/waverleysoftware/model/Panel.java:6]: I did it in the  **5-lombok-assert-annotation** branch. Please, look at the changes.

File: src/test/java/com/waverleysoftware/AccessCodeTests.java (hunk 9 above)
Comment (mkgerasimenko) [reply src/test/java/com/waverleysoftware/AccessCodeTests.java:18]: I did it in the  **5-lombok-assert-annotation** branch. Please, look at the changes.

**Review by sskorol — APPROVED** [review]

**Comment by sskorol** [comment]
LGTM

**Pull request merged**